src/librustc_typeck/check/mod.rs

```diff
@@ -1038,11 +1038,11 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
 
     let ret_ty = fn_sig.output();
     fcx.require_type_is_sized(ret_ty, decl.output.span(), traits::SizedReturnType);
-    let ret_ty = fcx.instantiate_anon_types_from_return_value(fn_id, &ret_ty);
-    fcx.ret_coercion = Some(RefCell::new(CoerceMany::new(ret_ty)));
+    let revealed_ret_ty = fcx.instantiate_anon_types_from_return_value(fn_id, &ret_ty);
+    fcx.ret_coercion = Some(RefCell::new(CoerceMany::new(revealed_ret_ty)));
```

Comment:
Indeed the diff is confusing because the line that changes behavior isn't on the diff. What do you think of renaming both?

What I wonder is: which naming would've had a better chance of preventing the bug? I think the name `revealed_ret_ty` calls attention to private information is being used.

Reply:
I think the code that has to see the type "from the outside" being *here* is part of the problem - after all, this part of typeck is working on *the body* of the function being checked. cc @nikomatsakis

Reply:
That makes sense, what do you think of moving the check to the `check_item_type` pass, right [here](https://github.com/rust-lang/rust/blob/75a00490c964a9263bf9d74800bd232ea0f63e69/src/librustc_typeck/check/mod.rs#L1199). I can try it in this PR, if we're not in a rush to land because of the point release.